### PR TITLE
Make atm build using CMake!

### PR DESCRIPTION
**Pull request by jgfouca**
Switch atm component over to using CMake. I tried to mimic the original behavior as closely as possible.

Change list:
* Change FFLAGS_NOPT to be an *append* to the normal set of flags. This reduces duplication with FFLAGS
* Remove Depends files for obsolete/unsupported platforms
* Make CMake versions of remaining Depends files
* New bld_diff tool for comparing bld logs, necessary to validate this work
* Add ability to ask for CMake args (get_standard_cmake_args)
* CMake macros should NOT be looking at environment, just CMake vars
* Homme test needs to use GMAKE_J and do a verbose build
* build_model_thread can now use run_sub_or_cmd 
* case_setup will make both macro types (Make and CMake)
* Import a FindPetsc CMake module
* components/CMakeLists.txt will be the CMake file to configure builds of our components. It currently is only doing so for cam.
* Remove obsolete bld files from cam
* Convert cam buildlib to python
* Someone appears to have added their genf90'd files to cam a long time ago, remove them.
* Convert cosp to Cmake. Cosp files are now just part of libatm, not their own library. This means CMake can handle the circular dependencies.

Current status:
* Appears to work on melvin (e3sm_developer passing), sandiatoss3 (e3sm_integration (build-only)), cori-knl (e3sm_integration (build-only))
* Build times look roughly the same, CMake is slightly faster
* Need to run analyzer tool on several platforms to ensure that atm builds are equivalent to what they were before.

User impact:
* Since this is just the initial phase, user impacts are minimal
* cosp integration with atm is different, it will need to be managed through cosp/Cosp.cmake.
* When adding Depends files for machine/compiler-specific flag changes, both a Make and CMake version will be needed until all components are switched over.
* It is no-longer enough to have NETCDF settings in the environment, they need to be explicit in the Macros (config_compilers.xml).

Putting up for an early review due to the high-impact of this PR.

[BFB]

### CODE REVIEW

**Comment by jgfouca** [comment]
@jedwards4b ^

**Comment by jgfouca** [comment]
[cmake_conversion.pptx](https://github.com/E3SM-Project/E3SM/files/3174546/cmake_conversion.pptx)

Updates + fleshed-out plan for CMake conversion.

**Comment by jgfouca** [comment]
All, this is ready for review.

**Comment by jgfouca** [comment]
This appears to be working for melvin (e3sm_developer), sandiatoss3 and cori-knl (e3sm_integration).

**Comment by jgfouca** [comment]
@rljacob anything more you want me to test before I push to next?

**Comment by rljacob** [comment]
Try e3sm_atm_extra_coverage.

**Comment by jgfouca** [comment]
I'm a bit perplexed at the lack of reviews. This will be the biggest change to CIME since the pythonization of the scripts and the biggest change to the guts of atm in ages.

**Comment by jedwards4b** [comment]
I've gone over it a couple of times and don't have any concerns as long as the scripts_regression_tests still pass and it still works with our cam gmake build.

**Comment by rljacob** [comment]
Lack of reviews just shows how few people are invested in the build system.  If it still works, they're ok with.

**Comment by jgfouca** [comment]
Looks like there may be problems with this on cori-knl afterall. I tested the wrong commit.

**Comment by bartgol** [comment]
This is a big undertaking. I guess my biggest concern is the centralized handling of the components. I would like each component to have their own CMakeLists.txt file, to perform component-specific stuff. E.g., which file they want to compile, rather than glob everything from the root folder, or add component-specific flags. 
On the other hand, that's something that the components devs need to do. Assuming we get someone to take care of each component, we could probably change this later (meaning that once all components have their individual cmakelists file, we can remove a bunch of stuff from components/CMakeLists.txt).

By the way, components/CMakeLists.txt defines the project, so I'm guessing it's the main CMakeLists.txt file. Yet, it does not add any target (library or executable), so there must be another place where cmake creates the final exec (or an intermediate library that cime links to), right?

**Comment by bartgol** [comment]
Another suggestion could be to split the CMakeLists.txt file. 800+ lines of cmake is a lot. Perhaps we could relegate some lstuff (like finding library paths and includes) to cmake scripts (in a components/cmake folder, perhaps). It's hard to not lose track of the configuration step we're at when going through 800+ lines of cmake.

**Comment by jgfouca** [comment]
@bartgol , that is the goal in the future. This PR is just getting CMake in place with the most naive, 1:1 conversion of Make possible. This is, by no means, what the final system will look like. Looking at the powerpoints linked above, this PR represents step 1.

> By the way, components/CMakeLists.txt defines the project, so I'm guessing it's the main CMakeLists.txt file. Yet, it does not add any target (library or executable), so there must be another place where cmake creates the final exec (or an intermediate library that cime links to), right?

The targets are defined at the bottom:
```
if (COMPLIB)
  add_library(${COMPLIB})
  target_sources(${COMPLIB} PRIVATE ${SOURCES})

  # Subtle: In order for fortran dependency scanning to work, our CPPFPP/DEFS must be registered
  # as COMPILE_DEFINITIONS, not simple added via CMAKE_Fortran_Flags. Also, CPPDEFS *must*
  # be provided as a list, not a whitespace-separated string; otherwise, things get wonky.
  separate_arguments(CPPDEFS_LIST UNIX_COMMAND "${CPPDEFS}")
  target_compile_definitions(${COMPLIB} PRIVATE ${CPPDEFS_LIST})
  add_dependencies(${COMPLIB} genf90)
endif()

if (EXEC_SE)
  add_executable(${EXEC_SE})
  target_link_library(${EXEC_SE} ${COMPLIB} ${ULIBDEP} ${CSMSHARELIB} ${MCTLIBS} ${PIOLIB} ${GPTLLIB} ${CLIBS} ${ULIBS} ${SLIBS} ${MLIBS} ${F90_LDFLAGS})
endif()
```

**Comment by bartgol** [comment]
Ah, I didn't see the ppt you posted. That makes sense.

File: components/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,840 @@
...
+# Note: Using absolute paths seems to wreck CMake's ability to do a dep analysis on fortran
+# sources and compile them in the right order.
+#
+# One additional subtley is that, when mkSrcfiles found multiple files with the same basename,
+# only the one found first gets compiled.
+
+#execute_process(COMMAND ${CASETOOLS}/mkSrcfiles)
+#file(STRINGS ${CMAKE_BINARY_DIR}/Srcfiles SOURCES)
+
+set(SRCROOT_REL "${CIMEROOT}/..")
```
Comment (ambrad) [inline components/CMakeLists.txt]: This might be factored out as a function (not a macro).

File: components/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+        list(APPEND SOURCES ${MATCH})
+        list(APPEND BASENAME_SET ${BASENAME})
+      else()
+        message(WARNING "Skipping repeated base filename ${BASENAME} for ${MATCH}")
+      endif()
+    endforeach()
+  endif()
+endforeach()
+
+foreach(SOURCE_FILE IN LISTS SOURCES)
```
Comment (ambrad) [inline components/CMakeLists.txt]: This too.

File: components/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+# so this needs to be added after creating CLIBS above
+set(CSMSHARELIB "${INSTALL_SHAREDPATH}/${COMP_INTERFACE}/${ESMFDIR}/${NINST_VALUE}/lib/libcsm_share.a")
+set(ULIBDEP "${ULIBDEP} ${CSMSHARELIB}")
+
+# Things may have been added to flags and we need to remove PIO stuff
+set(CMAKE_Fortran_FLAGS "${FFLAGS} ${INCLDIR}")
+set(CMAKE_C_FLAGS "${CFLAGS} ${INCLDIR}")
+set(CMAKE_CXX_FLAGS "${CXXFLAGS} ${INCLDIR}")
+
+# do the necessary genf90s
```
Comment (ambrad) [inline components/CMakeLists.txt]: And this.

File: cime/scripts/lib/CIME/build.py [hunk 4]
```diff
@@ -498,7 +519,7 @@ def _case_build_impl(caseroot, case, sharedlib_only, model_only, buildlist,
             os.environ["USE_KOKKOS"] = "TRUE"
```
Comment (ambrad) [inline cime/scripts/lib/CIME/build.py]: You might merge master into this branch and test again before merging this PR to next. This PR branches from master _before_ anything in master was building Kokkos on all machines. Now that PR #2896 is in master, all platforms are building Kokkos. On line 518 of this file (for example; there are other spots, too), only preqx_kokkos triggers Kokkos, whereas in master theta-l does, too, and theta-l builds in all nightlies, I believe.

Comment (jgfouca) [reply cime/scripts/lib/CIME/build.py]: Will do.

Comment (ambrad) [reply cime/scripts/lib/CIME/build.py]: I just remembered one detail. Right now there are no coupled theta-l tests. That means Kokkos is being build only as part of the test `HOMME_P24.f19_g16_rx1.A`. So to test the full Kokkos capability (both standalone Homme build, which is tested currently, and the CIME build, which currently is not), build a case with `./xmlchange CAM_TARGET=theta-l`.

Comment (jgfouca) [reply cime/scripts/lib/CIME/build.py]: Ok

**Comment by AaronDonahue** [comment]
@jgfouca , I decided to test this branch on the LC machines just to see if the changes have any impact on our build there.  The only thing I noted was that we have to add a `module load` of cmake to our machine file.  Could I add that to this branch directly so that once this PR goes in LC machines will be up and running without a separate PR?

Also, I took a look at how other machines are handling the cmake load to get an example and noticed that a number of the machine configs on this branch are loading an older version of cmake then is required.  If I'm not mistaken we need a version of 3.10 or greater.  I think for example cori-haswell which appears to be loading ver 3.3.2.  Just wanted to bring it up in case a simple fix to these will avoid headaches in the future.

**Comment by jgfouca** [comment]
@AaronDonahue , thank you for helping me test this PR. Yes, please add cmake updates to this branch. I had already updated cori but forgot to push, I'll do that now. Anvil also needs an update, but it only has 3.9, so I may need to reduce the over requirement to 3.9.

**Comment by AaronDonahue** [comment]
great, I'll add the changes and make the push

**Comment by jgfouca** [comment]
@AaronDonahue , it looks like I did push the cori update. You may be looking at a stale branch HEAD.

**Comment by jgfouca** [comment]
Quick update: I'm nearly complete with testing on all the E3SM_baseline machines and the E3SM_Machine_Coverage machines. I'd like to do an upstream merge tomorrow and integrate the recent change to scripts/Tools/Makefile into the CMakeLists file. After a bit of sanity checking, assuming all goes well, I'd like to integrate this into next sometime tomorrow.

**Comment by jgfouca** [comment]
I think this is ready. Any objections to merging it to next?

**Comment by rljacob** [comment]
Not on the day before a 3-day weekend.

**Comment by rljacob** [comment]
After further discussion: would be good to allow 3 days for all machines to return.   So no objections here.

**Comment by wlin7** [comment]
@jgfouca , I did a pair of tests for an F-case with COSP, one with this branch and the other with the master of 05/20/19. The build and run with this branch went smoothly. However, the results are not BFB. I don't see any non-BFB PR since the point that this branch was based off. Do you expect any non-BFB?

**Comment by rljacob** [comment]
It should be BFB and tests over the weekend passed.  But COSP was a little tricky so maybe you have a case we're not testing.

**Comment by jgfouca** [comment]
Looks like it works.

**Comment by jgfouca** [comment]
@wlin7 , can you elaborate? What machine, case did you test?

**Comment by wlin7** [comment]
@jgfouca , the runs were done on cori-knl. 

The cases were created with " -compset FC5AV1C-04P2 -res ne30_oECv3".
cosp is enabled by ./xmlchange CAM_CONFIG_OPTS="-cosp" --append.

One with hash ef9e1b9f38297551baeac666a3f51642636d88f6, master of 05/20.
The other with jgfouca/cam/cmakify, updated thru 05/20.

Differences in atm global statistics started by step 2. COSP may not have caused it as it does not feed back to atm model. I don't see any non-BFB PR in master betwen May 1 nnd May 20.

**Comment by jgfouca** [comment]
@wlin7 , ok, i'm trying to compare baselines with:

```
./create_test SMS.ne30_oECv3.FC5AV1C-04P2.cori-knl_intel.cam-cosplite -c -b fouc_cmake --wait
```

**Comment by wlin7** [comment]
@jgfouca , nice you are doing a formal test against the baseline. BTW,  it has been verified that COSP has no impact. I repeated the two cases w/o COSP, same non-BFB behavior, while each being BFB with its COSP counterparts.

**Comment by jgfouca** [comment]
@wlin7 , I'm unable to reproduce the diff you saw. What commit did you use to produce the baseline?

**Comment by jgfouca** [comment]
I think this is ready to merge.

**Comment by wlin7** [comment]
@jgfouca , I didn't follow a standard test procedure. In terms of baseline, I used hash ef9e1b9f38297551baeac666a3f51642636d88f6, following the merged of PR #2911.

**Comment by wlin7** [comment]
@jgfouca , all is good now. I did the same test using the current master with this PR merged. It is BFB with the master I tested earlier. The problem must be because I was using this PR up to only May 20 (4fe4d2620bce900b6eb46bcb79288eb60cc3d4a6). Certain commit/merge to this branch after May 20 makes a difference, but I am not able to isolate. It does not matter, though.

**Comment by jgfouca** [comment]
@wlin7 , yeah I did an upstream merge to the branch sometime after you made your original baseline. Glad things are good now.

**Pull request merged**